### PR TITLE
DateMath: Interactivity

### PR DESCRIPTION
**Pull request by GuiltyDolphin**
This adds interactivity to the DateMath Goodie through JavaScript.

_Hide your eyes._
### To-Do
- [x] Use a single line for the amount input, in the form `Years: X Months: X etc.` with support for changing the operation from addition to subtraction.

---

IA Page: https://duck.co/ia/view/date_math

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by GuiltyDolphin** [comment]
@mintsoft Want to have a look over this? I think the functionality is all there; just a question of getting it to look nice.

I'll have another look pretty soon to make sure I've removed everything that should be, and clean up anything that requires it.

**Comment by mintsoft** [comment]
@GuiltyDolphin certainly can

**Comment by mintsoft** [comment]
Hmm so I'm a smidge confused by the GUI at first; I've just played with it and it made more sense once I could see how it reacted.

@abeyang Any design thoughts here?

**Comment by GuiltyDolphin** [comment]
@mintsoft Glad it made sense (eventually)! I'm no designer so any design review is appreciated.

**Comment by mintsoft** [comment]
@GuiltyDolphin Immediate things that I've noticed are that I think some labels might help?
* It'd be nice to highlight the "start date"
* The input boxes on the start bit are a little difficult to tell that you can edit them, because some bits you can, some you can't
* Personally I dislike month first date formats; I'd prefer like `Monday 25th April 2016 21:17:05`

**Comment by mintsoft** [comment]
Also it doesn't give me the answer when I query? I have to tinker with it to get the answer out?

**Comment by GuiltyDolphin** [comment]
Oh lord @mintsoft yeah - forgot to turn that on!

**Comment by GuiltyDolphin** [comment]
@mintsoft Sure - I was using a random format suggested by `moment.js`. Probably best to go with the general 'DuckDuckGo' format (with weekday).

**Comment by mintsoft** [comment]
Ah somehow I knew it'd be `moment.js`'s fault

**Comment by GuiltyDolphin** [comment]
@mintsoft I've updated the date format and fixed the date not being initially calculated.

For the 'indication' that the fields are editable; I'm pretty happy with the current 'neutral' style for the date, so how about a [popout](https://duckduckgo.com/styleguide#popouts) or tooltip?

**Comment by mintsoft** [comment]
@GuiltyDolphin I think my confusion around the input boxes is because that line is small and you highlight things like the sunday it makes it look like it's just text. Maybe if you did ` user-select: none; ` (and all the prefixes for different browsers) it'd be more obvious? I dunno

**Comment by mintsoft** [comment]
Hmm maybe change them to `<input type='number' />` ? Then they'll have little spinners on them too; makes it a bit clearer when you click it

**Comment by GuiltyDolphin** [comment]
@mintsoft I had that (spinners) before - it was pretty ugly.

There is the indication when you hover over (a caret); but I guess we could make the highlighting a bit more prominent - I'll see how it looks.

**Comment by moollaza** [comment]
Just looking over this. My initial reaction is that there's too much JS here. It looks like a lot of what you've implemented can be handled by the browser:

- Changing the input `type` to number will and setting `max=`, `min=` will help a lot
- That should force only acceptable inputs, meaning no more validation required
- Our `btn` class should handle all the UI aspects of any buttons: https://duckduckgo.com/styleguide#buttons
- Our grid classes should help with the layout -- though I suspect our design team will have thoughts on the overall UI

File: share/goodie/date_math/date_math.js [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    "use strict";
+
+    DDH.date_math.build = function(ops) {
+
+        var saData = ops.data;
+        var modifier_order = [
+            'Second', 'Minute', 'Hour',
+            'Day', 'Week', 'Month', 'Year'
+        ];
+        function pad(text, padChar, padWidth) {
```
Comment (moollaza) [inline share/goodie/date_math/date_math.js]: Move private functions outside the `build` function definition

File: share/goodie/date_math/date_math.js [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                    $year = $dom.find('.input--year'),
+                    $startDate = $dom.find('.date--start'),
+                    $resultDate = $dom.find('.date--result'),
+                    $opTemplate = $dom.find('.template--op').clone(),
+                    $ops = $dom.find('.op--container'),
+                    $hour = $dom.find('.input--hour'),
+                    $minute = $dom.find('.input--minute'),
+                    $second = $dom.find('.input--second');
+
+                $dom.find('form').submit(function(e) {
```
Comment (moollaza) [inline share/goodie/date_math/date_math.js]: looks like you can remove this -- we're not using a `<form>`

File: share/goodie/date_math/date_math.js [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                DDG.require('moment.js', function() {
+                    function getDate() {
+                        var month = getMonth();
+                        var day = getDay();
+                        var year = getYear();
+                        var hour = $hour.val();
+                        var minute = $minute.val();
+                        var second = $second.val();
+                        var time = hour + ':' + minute + ':' + second;
+                        var date = moment(year + "-" + month + "-" + day +
```
Comment (moollaza) [inline share/goodie/date_math/date_math.js]: I would use `moment().set({'year': 2013, 'month': 3}).format(...);` or `moment().year(NNNN).month(NN).date(NN).format(.....)`

**Comment by GuiltyDolphin** [comment]
@moollaza 

https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2897/commits/34ca44f756e4d0514b577194dc43c1111e8d365d - I was using `number` with `min` and `max` before; but I think it was causing some display issues; I'll check it out again.

I am using `btn` - but there's some overriding going on that made it a little painful and I had to add some custom stuff... Some design advice/feedback would be very welcome!

Thanks for the review :smile:

File: share/goodie/date_math/date_math.js (hunk 2 above)
Comment (GuiltyDolphin) [inline share/goodie/date_math/date_math.js]: Yeah, I think I was before - there may be a few other legacy pieces, let me know if you spot any!

**Comment by GuiltyDolphin** [comment]
@moollaza BTW, which (specifically) were you thinking should be switched to `number`? The `day` is supposed to allow inputs like `2nd`, `3rd` etc., the amount ends up looking like:

screenshot-2016-06-02-21 12 42_598_47

The times get really screwed up as well.

Am I missing something? It feels like there should be a way to *not* have it screw everything up and just have the input type; though looking at other IAs they've used things like `tele` and `text` inputs to bypass this.

**Comment by GuiltyDolphin** [comment]
@moollaza Ah, playing around with styles a bit I think I might be able to replace some of the `text` input with `number`.

**Comment by GuiltyDolphin** [comment]
@moollaza Right, I've changed some of the fields to use `tel` and `number`, and I've removed a bunch of the custom validity checking; allowing the removal of a fair bit o' JS.

I think I preferred the old checks, but this should serve :+1:

In terms of functionality, I think this is there - I think the design needs touching up a 'little', but apart from that... This should be fairly complete.

Would it be possible to get this on Beta soonish?

**Comment by moollaza** [comment]
> Right, I've changed some of the fields to use tel and number, and I've removed a bunch of the custom validity checking; allowing the removal of a fair bit o' JS.

Hmm I'm not a fan of using `tel` for anything that isn't actually a **tel**ephone number -- that's just abusing HTML.

I plan to speak to the team about how we can better handle all these forms and inputs, and also handle DOM changes. There's no shortage of JS validation libs and I don't want us to have to implement our own checks in each IA. Similarly we should be able to bind all these inputs to model so we don't need to keep searching the DOM and modifying the `html()` and `val()` etc. That's just icky.

>The day is supposed to allow inputs like 2nd, 3rd etc., the amount ends up looking like:

Why do we need to handle inputs like "2nd", "3rd", etc?

**Comment by GuiltyDolphin** [comment]
@moollaza Uhh... We don't need to handle inputs like `2nd`, `3rd` etc., that was my bad - it just normalizes to that (also meaning that `number` doesn't work).

Whoops - yeah, we can just use 'text' for a bunch of these (as before); just with less checks.

**Comment by moollaza** [comment]
Updated on Beta.

File: share/goodie/date_math/content.handlebars [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        size='2' class="input--second frm__input" />
+    </span>
+  </span>
+  <span class="frm g ten">
+    <p class="frm__desc tx--11">ACTION</p>
+    <span class="btn ddgsi-plus input--op-sign" />
+  </span>
+  <span class="input--op g thirty">
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select gw">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: remove this `gw`

File: share/goodie/date_math/content.handlebars [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="g five" />
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: remove this `<span>`

File: share/goodie/date_math/content.handlebars [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<div class="gw">
+  <span class="input--date">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: add `sixty`

File: share/goodie/date_math/content.handlebars [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    <span class="frm g thirty">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: change to `sixty`

File: share/goodie/date_math/content.handlebars [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
+    </span>
+    <span class="frm input--time input--numeric g thirty">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: change to `forty`

File: share/goodie/date_math/content.handlebars [hunk 9]
```diff
@@ -0,0 +1,38 @@
+<div class="gw">
+  <span class="input--date">
+    <span class="frm g thirty">
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: add a CSS rule for this to have  `width: 100%`

File: share/goodie/date_math/content.handlebars [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  <span class="frm g ten">
+    <p class="frm__desc tx--11">ACTION</p>
+    <span class="btn ddgsi-plus input--op-sign" />
+  </span>
+  <span class="input--op g thirty">
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select gw">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="g five" />
+      <span class="frm__select g fifty">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: need to add css to set `height: 2em` might need `!important`

File: share/goodie/date_math/content.handlebars [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        <select class="input--op-type" />
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: same here, need to add css to set `height: 2em`

File: share/goodie/date_math/content.handlebars [hunk 12]
```diff
@@ -0,0 +1,38 @@
+<div class="gw">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: Give this a class like `.container`

add css for `margin-bottom: 1em`

**Comment by moollaza** [comment]
@GuiltyDolphin this is looking amazing 👍 💯  Nicely done.

I've left some comments about CSS fixes that should be made. The Pikaday css is slightly broken causing the modal to have no background, but I've already made a simple internal PR. Once that's deployed to beta the pikaday selector will work just fine.

The end result of my CSS looks like this:

jan_1_2012_plus_32_days_at_duckduckgo

File: share/goodie/date_math/content.handlebars (hunk 10 above)
Comment (GuiltyDolphin) [inline share/goodie/date_math/content.handlebars]: @moollaza Hmm, tried setting the style via CSS (and adding IDs, and a bunch more classes to try and improve chances of being selected) but no luck (with or without `!important`); works if I use `style="height:2em"` though.

**Comment by GuiltyDolphin** [comment]
@moollaza Thanks!

I've updated the design per your suggestions - it looks better now but my version appears to have a couple of issues still (not sure if this is just me). Some of the external CSS *refused* to be applied to the elements so I've had to use `style="..."` on some to *force* the changes.

**Comment by tagawa** [comment]
I just installed the updated version to beta: https://beta.duckduckgo.com/?q=Jan%201%202012%20plus%2032%20days

datetime_ia_screenshot

Looks like we may need to change the class or style for the Result line, and the length of the calendar field. Would using `type=date` for the input help there?

File: share/goodie/date_math/content.handlebars [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  <span class="input--date sixty">
+    <span class="frm g sixty">
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
+    </span>
+    <span class="frm input--time input--numeric g forty">
+      <p class="frm__desc tx--11">TIME</p>
+      <input type="text" maxlength='2' pattern='^(2[0-3]|[0-1]?[0-9])$'
+        size='2' class="input--hour frm__input" />
+      <span class="sep--time">:</span>
```
Comment (tagawa) [inline share/goodie/date_math/content.handlebars]: Would it be easier to use `<input type="number" min=0 max=23 class="input--hour frm__input" />` here, and similar for the subsequent fields?

Comment (GuiltyDolphin) [inline share/goodie/date_math/content.handlebars]: @tagawa I've had a few issues with the `number` input type, the most pressing in this case being that you cannot (easily) normalize the format to two digits (e.g., `7` -> `07`); so not really :/

Comment (tagawa) [inline share/goodie/date_math/content.handlebars]: @GuiltyDolphin Understood. Please ignore my comment.

File: share/goodie/date_math/content.handlebars [hunk 14]
```diff
@@ -0,0 +1,37 @@
+<div class="gw .container" style="margin-bottom:1em">
+  <span class="input--date sixty">
+    <span class="frm g sixty">
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
```
Comment (tagawa) [inline share/goodie/date_math/content.handlebars]: Would `type="date"` be easier here? It would avoid the text truncation issue if so.

Comment (GuiltyDolphin) [inline share/goodie/date_math/content.handlebars]: @tagawa I don't get any wrapping locally (though the rest of the display is a bit off):

screenshot-2016-07-28-10 45 45_876_268

Though switching it to `date` doesn't seem to make a difference, so I'll try that.

**Comment by GuiltyDolphin** [comment]
@tagawa I've fixed the result line color and updated the input field to `date` - with any luck that'll fix it on beta (though I didn't notice any difference locally).

**Comment by tagawa** [comment]
Thanks @GuiltyDolphin. The dark theme (and other themes) look good now, and personally I find the date input easy to use. A couple of points:

1. The three Time boxes are wrapping in Firefox for me. Safari and Vivaldi are fine.

2. The output date is not showing in Vivaldi (and I assume Opera, Chrome, etc.) because the date input is being filled with an incorrect format - it expects "yyyy-MM-dd". Perhaps it's OK if we use the browser-accepted format for the input date, because the output date is human-friendly and that's the most important (from the user's point of view).

date-calculator-screenshots

**Comment by GuiltyDolphin** [comment]
@tagawa I'll just revert the changes to the date input - they can't complain if it is a text field.

Hmm... Locally the boxes wrap for me, but on Beta it's fine, weird... (I'm on FF).

File: share/goodie/date_math/content.handlebars (hunk 14 above)
Comment (GuiltyDolphin) [inline share/goodie/date_math/content.handlebars]: Reverted due to browser issues.

File: share/goodie/date_math/content.handlebars [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  </span>
+  <span class="frm g ten">
+    <p class="frm__desc tx--11">ACTION</p>
+    <span class="btn ddgsi-plus input--op-sign" />
+  </span>
+  <span class="input--op g thirty">
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="frm__select g fifty" style="height:2em">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: use `!important` to override CSS -- inline styles shouldn't be necessary -- or give these a new class and create a more specific selector

File: share/goodie/date_math/content.handlebars [hunk 16]
```diff
@@ -0,0 +1,37 @@
+<div class="gw .container" style="margin-bottom:1em">
+  <span class="input--date sixty">
+    <span class="frm g sixty">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: You can remove this `<span>`

File: share/goodie/date_math/content.handlebars [hunk 17]
```diff
@@ -0,0 +1,37 @@
+<div class="gw .container" style="margin-bottom:1em">
+  <span class="input--date sixty">
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: Whoops, meant you can remove this one, or at least the `sixty`. You can control the sizes of the other elements better to allow more space for the `<select>` at the end

Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: I switched:
- the DATE input to `sixty`
- the `date--start` to `forty`
- `input--time` to `quarter`
- `input--op` to `quarter`

The HTML I ended up with looks like:

```html
<div class="gw container" style="margin-bottom:1em">
    <p class="frm__desc tx--11">
        <span class="frm g forty">DATE</span>
    </p>
    <span class="frm g forty">
        <input type="text" class="frm__input date--start" />
    </span>

    <p class="frm__desc tx--11">
        <span class="frm input--time input--numeric g quarter">TIME</span>
    </p>
    <span class="frm input--time input--numeric g quarter">
        <input type="text" maxlength="2" pattern="^(2[0-3]|[0-1]?[0-9])$" size="2" class="input--hour frm__input" />
        <span class="sep--time">:</span>
        <input type="text" maxlength="2" pattern="^([0-5][0-9]|[0-9])$" size="2" class="input--minute frm__input" />
        <span class="sep--time">:</span>
        <input type="text" maxlength="2" pattern="^([0-5][0-9]|[0-9])$" size="2" class="input--second frm__input" />
    </span>

    <p class="frm__desc tx--11">
        <span class="frm g ten">ACTION</span>
    </p>

    <p class="frm__desc pull--left tx--11">
        <span class="input--op g quarter" style="">DURATION</span>
    </p>
    <span class="input--op g quarter" style="">
        <span class="frm op--select">
            <input type="number" class= "frm__input input--op-amt forty" />
            <span class="frm__select sixty" style="height:2em">
                <select class="input--op-type" style="height:2em">
                    <option value="second">Seconds</option>
                    <option value="minute">Minutes</option>
                    <option value="hour">Hours</option>
                    <option value="day">Days</option>
                    <option value="week">Weeks</option>
                    <option value="month">Months</option>
                    <option value="year">Years</option>
                </select>
            </span>
        </span>
    </span>
</div>
```

**Comment by GuiltyDolphin** [comment]
@moollaza Uhh, that appears to be missing the sign toggle? 

On Midori it takes it from:



To:



(I've played around with it a bit).

I'm not 100% sure what the issue is - on FF it looks screwed up locally, but on Beta it's fine. Could it just be something to do with slightly different styles on Beta and what DuckPAN receives?

File: share/goodie/date_math/content.handlebars (hunk 15 above)
Comment (GuiltyDolphin) [inline share/goodie/date_math/content.handlebars]: @moollaza See my comment above - I've tried using CSS with multiple classes + id + tag etc. (with `!important`) to no avail - unless I'm missing something obvious it looks like I can't override them?

**Comment by moollaza** [comment]
> Uhh, that appears to be missing the sign toggle?

@GuiltyDolphin looks like your pictures didn't upload in that last comment?

> on FF it looks screwed up locally, but on Beta it's fine.

I'll take a look at FF...

> Could it just be something to do with slightly different styles on Beta and what DuckPAN receives?

Shouldn't be. If any, it would only be minor and temporary (once Beta is updated internally). AFAIK we haven't made any changes to CSS lately.

File: share/goodie/date_math/content.handlebars (hunk 15 above)
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: `.zci--date_math .frm__select.g` seems to work?

File: share/goodie/date_math/content.handlebars [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  <span class="frm g ten">
+    <p class="frm__desc tx--11">ACTION</p>
+    <span class="btn ddgsi-plus input--op-sign" />
+  </span>
+  <span class="input--op g thirty">
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="frm__select g fifty" style="height:2em">
+        <select class="input--op-type" style="height:2em"/>
```
Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: can use `.frm__select select.input--op-type` with `!important` here.

Our existing style has `!important` so we need to make our selector more specific first, then force it with `!important`

Comment (moollaza) [inline share/goodie/date_math/content.handlebars]: You'll want to also remove the padding on the `select` the `span` already adds some, forcing the `seconds` option to overlap with the dropdown

**Comment by moollaza** [comment]
@GuiltyDolphin if you reduce the padding-left on the `.time--input .frm__input` then it should prevent the wrapping on FF. 

It may be wrapping due to some kind of calculation bug in FF... Not sure why the math isn't adding up there.

**Comment by GuiltyDolphin** [comment]
@moollaza Everything just seems... Bigger, locally:

screenshot-2016-08-06-18 35 19_857_220

screenshot-2016-08-06-18 35 33_870_274

(Beta/Local)

Changing anything just seems to screw it up more (locally); @tagawa was that screenshot with the wrapping on FF on beta or local?

**Comment by mintsoft** [comment]
I'm going to close this for now until we decide what we want to do with it, feel free to re-open if you disagree

**Pull request closed without merging**